Sources/NIO/EventLoop.swift

```diff
@@ -1104,6 +1112,27 @@ public final class MultiThreadedEventLoopGroup: EventLoopGroup {
         // our shutdown signaling, and then do our cleanup once the DispatchQueue is empty.
         let g = DispatchGroup()
         let q = DispatchQueue(label: "nio.shutdownGracefullyQueue", target: queue)
+        do {
+            self.shutdownLock.lock()
+            defer { shutdownLock.unlock() }
```

Comment:
The problem is that we actually `return` from the function, which is not possible from within `self.shutdownLock.withLockVoid`. If you feel strongly, I could make it return a `Bool` or something that signals whether or not it should return.

Reply:
If we return, then use `Lock.withLock`.

Reply:
No, I mean return from the surrounding function, but only if we are already stopping. But we only know that once we check the status, for which we acquire the lock. But we also may have to modify the status before we return, so we have to keep the lock until then. So the only way I see to make it work with `withLock` is to return a value from the `withLock` that indicates whether or not we should return from the function, which to me seems worse than just keeping it as it is, but I can make that change if that‘s the preferred style.

Reply:
We should return a value that indicates whether we need to return from the function, IMO. Restricting the scope of locks is good, as is clearly delimiting their end point, so I think the refactor is worth it.

Of course, the other way to make it work is to refactor the body of this code into a separate inner function that is called from only the `.running` case. That allows you to keep the extra lock scope, though I don't think it's necessary here.

Reply:
Ok, I'll make that change then.

I think the separate function wouldn't help, because that would extend the scope of the lock to the whole function, instead of just the status checking.

Reply:
Oh yeah, please remove the `defer` statement.

Reply:
😅 sorry about that. Interestingly the tests succeeded on my machine 🤨

Reply:
I think double lock release isn't guaranteed to do anything good or bad. Good job Linux was paying attention to the mutex state!